References
==========

How to Cite Us?
---------------

If you use FANTASIA in your work, please cite:

* Martínez-Redondo GI, Pérez-Canales FM, Carbonetto B, Fernández JM,
  Barrios-Núñez I, Vázquez-Valls M, Cases I, Rojas AM, Fernández R. (2025).
  FANTASIA leverages language models to decode the functional dark proteome
  across the animal tree of life. *Communications Biology* 8, 1227.
  doi: `10.1038/s42003-025-08651-2 <https://doi.org/10.1038/s42003-025-08651-2>`_.


Software and Libraries
----------------------

* PostgreSQL Global Development Group. PostgreSQL. Available at:
  `PostgreSQL <https://www.postgresql.org/>`_.

* pgvector: Open-source vector similarity search for Postgres. Available at:
  `pgvector <https://github.com/pgvector/pgvector>`_.

* RabbitMQ: Open-source message broker. Available at:
  `RabbitMQ <https://www.rabbitmq.com/>`_.

* McKinney W. (2010). Data Structures for Statistical Computing in Python.
  *Proceedings of the 9th Python in Science Conference*, 56–61. [pandas]

* Polars: A lightning-fast DataFrame library. Available at:
  `Polars <https://www.pola.rs/>`_.

* Klopfenstein DV, Zhang L, Pedersen BS, Ramírez F, Warwick Vesztrocy A, Naldi A,
  Mungall CJ, Yunes JM, Botvinnik O, Weigel M, Dampier W, Dessimoz C, Flick P,
  Tang H. (2018). GOATOOLS: A Python library for Gene Ontology analyses.
  *Scientific Reports* 8: 10872.
  doi: `10.1038/s41598-018-28948-z <https://doi.org/10.1038/s41598-018-28948-z>`_.

* Daily J. (2016). Parasail: SIMD C library for global, semi-global, and local
  pairwise sequence alignments. *BMC Bioinformatics* 17: 81.
  doi: `10.1186/s12859-016-0930-z <https://doi.org/10.1186/s12859-016-0930-z>`_.

* Huerta-Cepas J, Serra F, Bork P. (2016). ETE 3: Reconstruction, Analysis, and
  Visualization of Phylogenomic Data. *Molecular Biology and Evolution* 33(6):
  1635–1638.
  doi: `10.1093/molbev/msw046 <https://doi.org/10.1093/molbev/msw046>`_.
  PMID: 26921390; PMCID: PMC4868116.

* Virtanen P, Gommers R, Oliphant TE, et al. (2020). SciPy 1.0: Fundamental
  Algorithms for Scientific Computing in Python. *Nature Methods* 17: 261–272.
  doi: `10.1038/s41592-019-0686-2 <https://doi.org/10.1038/s41592-019-0686-2>`_.

* Paszke A, Gross S, Massa F, Lerer A, Bradbury J, Chanan G, Killeen T,
  Lin Z, Gimelshein N, Antiga L, Desmaison A, Köpf A, Yang E, DeVito Z,
  Raison M, Tejani A, Chilamkurthy S, Steiner B, Fang L, Bai J, Chintala S.
  (2019). PyTorch: An Imperative Style, High-Performance Deep Learning Library.
  *arXiv preprint* arXiv:1912.01703.
  `https://arxiv.org/abs/1912.01703 <https://arxiv.org/abs/1912.01703>`_.

* Wolf T, Debut L, Sanh V, Chaumond J, Delangue C, Moi A, Cistac P, Rault T,
  Louf R, Funtowicz M, Davison J, Shleifer S, von Platen P, Ma C, Jernite Y,
  Plu J, Xu C, Le Scao T, Gugger S, Drame M, Lhoest Q, Rush AM. (2020).
  HuggingFace's Transformers: State-of-the-art Natural Language Processing.
  *arXiv preprint* arXiv:1910.03771.
  `https://arxiv.org/abs/1910.03771 <https://arxiv.org/abs/1910.03771>`_.


Scientific References
---------------------

* Martínez-Redondo GI, Pérez-Canales FM, Carbonetto B, Fernández JM,
  Barrios-Núñez I, Vázquez-Valls M, Cases I, Rojas AM, Fernández R. (2025).
  FANTASIA leverages language models to decode the functional dark proteome
  across the animal tree of life. *Communications Biology* 8, 1227.
  doi: `10.1038/s42003-025-08651-2 <https://doi.org/10.1038/s42003-025-08651-2>`_.

* Barrios-Núñez I, Martínez-Redondo GI, Medina-Burgos P, Cases I,
  Fernández R, Rojas AM. (2024). Decoding functional proteome information in
  model organisms using protein language models. *NAR Genomics and
  Bioinformatics* 6(3): lqae078.
  doi: `10.1093/nargab/lqae078 <https://doi.org/10.1093/nargab/lqae078>`_.

* Littmann M, Heinzinger M, Dallago C, Olenyi T, Rost B. (2021). Embeddings
  from deep learning transfer GO annotations beyond homology. *Scientific
  Reports* 11: 1160.
  doi: `10.1038/s41598-020-80786-0 <https://doi.org/10.1038/s41598-020-80786-0>`_.

* Dallago C, Schütze K, Heinzinger M, Olenyi T, Littmann M, Lu AX, Yang KK,
  Min S, Yoon S, Morton JT, Rost B. (2021). Learned embeddings from deep
  learning to visualize and predict protein sets. *Current Protocols* 1: e113.
  doi: `10.1002/cpz1.113 <https://doi.org/10.1002/cpz1.113>`_.
